hal/halx86/smp/i386/spinup.c

```diff
@@ -0,0 +1,132 @@
+/*
+ * PROJECT:     ReactOS Kernel
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     i386 Application Processor (AP) spinup setup
+ * COPYRIGHT:   Copyright 2021 Justin Miller <justinmiller100@gmail.com>
+ *              Copyright 2021 Victor Perevertkin <victor.perevertkin@reactos.org>
+ */
+
+/* INCLUDES ******************************************************************/
+
+#include <hal.h>
+#include <smp.h>
+
+#define NDEBUG
+#include <debug.h>
+
+/* GLOBALS *******************************************************************/
+
+extern PPROCESSOR_IDENTITY HalpProcessorIdentity;
+extern PHYSICAL_ADDRESS HalpLowStubPhysicalAddress;
+extern PVOID HalpLowStub;
+
+// The data necessary for a boot (stored inside HalpLowStub)
+extern PVOID HalpAPEntry16;
+extern PVOID HalpAPEntryData;
+extern PVOID HalpAPEntry32;
+extern PVOID HalpAPEntry16End;
+extern HALP_APIC_INFO_TABLE HalpApicInfoTable;
+
+ULONG HalpStartedProcessorCount = 1;
+
+#ifndef Add2Ptr
+#define Add2Ptr(P,I) ((PVOID)((PUCHAR)(P) + (I)))
+#endif
+#ifndef PtrOffset
+#define PtrOffset(B,O) ((ULONG)((ULONG_PTR)(O) - (ULONG_PTR)(B)))
+#endif
+
+typedef struct _AP_ENTRY_DATA
+{
+    ULONG Jump32Offset;
+    ULONG Jump32Segment;
+    PVOID SelfPtr;
+    ULONG PageTableRoot;
+    PKPROCESSOR_STATE ProcessorState;
+    KDESCRIPTOR Gdtr;
+} AP_ENTRY_DATA, *PAP_ENTRY_DATA;
+
+typedef struct _AP_SETUP_STACK
+{
+    PVOID ReturnAddr;
+    PVOID KxLoaderBlock;
+} AP_SETUP_STACK, *PAP_SETUP_STACK;
+
+/* FUNCTIONS *****************************************************************/
+
+static
+ULONG
+HalpSetupTemporaryMappings(
+    _In_ PKPROCESSOR_STATE ProcessorState)
+{
+    PMMPDE RootPageTable = ExAllocatePoolZero(NonPagedPool, PAGE_SIZE, TAG_HAL);
+    PMMPDE LowMapPde = Add2Ptr(HalpLowStub, PAGE_SIZE);
+    PMMPTE LowStubPte = MiAddressToPte(HalpLowStub);
+    PHYSICAL_ADDRESS PhysicalAddress;
+    ULONG StartPti;
+
+    /* Copy current mappings */
+    RtlCopyMemory(RootPageTable, MiAddressToPde(NULL), PAGE_SIZE);
+
+    /* Set up low PDE */
+    PhysicalAddress = MmGetPhysicalAddress(LowMapPde);
+    RootPageTable[0].u.Hard.PageFrameNumber = PhysicalAddress.QuadPart >> PAGE_SHIFT;
+    RootPageTable[0].u.Hard.Valid = 1;
+    RootPageTable[0].u.Hard.Write = 1;
+
+    /* Copy low stub PTEs */
+    StartPti = MiAddressToPteOffset(HalpLowStubPhysicalAddress.QuadPart);
+    ASSERT(StartPti + 10 < 1024);
+    for (ULONG i = 0; i < 10; i++)
+    {
+        LowMapPde[StartPti + i] = LowStubPte[i];
+    }
+
+    PhysicalAddress = MmGetPhysicalAddress(RootPageTable);
+    ASSERT(PhysicalAddress.QuadPart < 0x100000000);
+    return (ULONG)PhysicalAddress.QuadPart;
+}
+
+BOOLEAN
+NTAPI
+HalStartNextProcessor(
+    _In_ PLOADER_PARAMETER_BLOCK LoaderBlock,
+    _In_ PKPROCESSOR_STATE ProcessorState)
+{
+    /* Write KeLoaderBlock into Stack */
+    ProcessorState->ContextFrame.Esp = (ULONG)((ULONG_PTR)ProcessorState->ContextFrame.Esp - sizeof(AP_SETUP_STACK));
```

Comment:
With my previous comment (about the parameter(s) of the kernel entry-point to be pushed onto the stack): it would be nice for that HalStartNextProcessor to _**NOT**_ modify the contents of the ProcessorState buffer, so that, the caller (i.e. KeStartAllProcessors) of HalStartNextProcessor could be sure that the ProcessorState buffer is not modified, and it could be initialized (mostly) only once, outside of its while(TRUE) loop.